cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuthorizationEnforcer.java

```diff
@@ -53,9 +54,38 @@
    */
   void enforce(EntityId entity, Principal principal, Set<Action> actions) throws Exception;
 
+  /**
+   * Enforces authorization for the specified {@link Principal} for at least one of the specified {@link Action actions}
+   * on the specified {@link EntityId}.
+   *
+   * @param entity the {@link EntityId} on which authorization is to be enforced
+   * @param principal the {@link Principal} that performs the actions
+   * @param actions the {@link Action actions} being performed, of which at least one must be allowed
+   * @throws UnauthorizedException if the principal is not authorized to perform any of the actions on the entity
+   * @throws Exception if any other errors occurred while performing the authorization enforcement check
+   */
+  //void enforceAtLeastOneAction(EntityId entity, Principal principal, Set<Action> actions) throws Exception;
+
+  /**
+   * Checks whether a single {@link EntityId} is visible to the specified {@link Principal}.
+   * An entity is visible to a principal if the principal has any privileges on the entity, or any of its descendants.
+   * However, visibility check behavior can be overwritten at the authorization extension level.
+   *
+   * @param entityId the entity on which the visibility check is to be performed
+   * @param principal the principal to check the visibility for
+   * @throws UnauthorizedException if the entity is not visible to the principal
+   * @throws Exception if any errors occurred while performing the check
+   */
+  default void isVisible(EntityId entityId, Principal principal) throws Exception {
```

Comment:
btw, any reason we need a default implementation here?

Reply:
We need it to be default for backwards compatibility. Additionally, we need it in the first place due to being unable to pass UnauthorizedExceptions consistently through the other isVisible method.